src/async.rs

```diff
@@ -0,0 +1,228 @@
+    fn sleep(&mut self, duration: Duration) -> Self::Future {
+        reactor::Timeout::new(duration, self).into_future().flatten()
+    }
+}
+
+enum RetryState<S, A> where S: Sleep, A: IntoFuture {
+    Running(A::Future),
+    Sleeping(S::Future)
+}
+
+/// Future that drives multiple attempts at an operation.
+pub struct RetryFuture<S, I, O, A> where S: Sleep, I: IntoIterator<Item=Duration>, A: IntoFuture, O: FnMut() -> A {
+    strategy: I::IntoIter,
+    state: RetryState<S, A>,
+    operation: O,
+    sleep: S,
+    total_delay: Duration,
+    tries: u64
+}
+
+impl<S, I, O, A> fmt::Debug for RetryFuture<S, I, O, A> where S: Sleep, I: IntoIterator<Item=Duration>, A: IntoFuture, O: FnMut() -> A {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "RetryFuture {{ total_delay: {:?}, tries: {:?} }}", self.total_delay, self.tries)
+    }
+}
+
+/// Retry the given operation asynchronously until it succeeds, or until the given Duration iterator ends.
+pub fn retry<S, I, O, A>(sleep: S, iterable: I, operation: O) -> RetryFuture<S, I, O, A> where S: Sleep, I: IntoIterator<Item=Duration>, A: IntoFuture, O: FnMut() -> A {
+    RetryFuture::spawn(sleep, iterable, operation)
+}
+
+impl<S, I, O, A> RetryFuture<S, I, O, A> where S: Sleep, I: IntoIterator<Item=Duration>, A: IntoFuture, O: FnMut() -> A {
+    fn spawn(sleep: S, iterable: I, mut operation: O) -> RetryFuture<S, I, O, A> {
+        RetryFuture {
+            strategy: iterable.into_iter(),
+            state: RetryState::Running(operation().into_future()),
+            operation: operation,
+            sleep: sleep,
+            total_delay: Duration::default(),
+            tries: 1
+        }
+    }
+
+    fn attempt(&mut self) -> Poll<A::Item, RetryError<A::Error>> {
+        let future = (self.operation)().into_future();
+        self.state = RetryState::Running(future);
+        return self.poll();
+    }
+
+    fn retry(&mut self, err: A::Error) -> Poll<A::Item, RetryError<A::Error>> {
+        match self.strategy.next() {
+            None => Err(RetryError::Operation{
+                error: err,
+                total_delay: self.total_delay,
+                tries: self.tries
+            }),
+            Some(duration) => {
+                self.total_delay += duration;
+                self.tries += 1;
+                let future = self.sleep.sleep(duration);
+                self.state = RetryState::Sleeping(future);
+                return self.poll();
+            }
+        }
+    }
+}
+
+impl<S, I, O, A> Future for RetryFuture<S, I, O, A> where S: Sleep, I: IntoIterator<Item=Duration>, A: IntoFuture, O: FnMut() -> A {
+    type Item = A::Item;
+    type Error = RetryError<A::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let result = match self.state {
+            RetryState::Running(ref mut future) =>
+                Either::Left(future.poll()),
```

Comment:
If we can use the `Either` from the futures crate, let's do that! Otherwise, leave it as is. We can always remove it later if we figure out a way around it.